File: apps/windows/TeslaSync.App/feature-views/ChangesPanel/ChangesPanel.Model.cs
using System.Globalization;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews.FeatureFlags;

/// <summary>
/// The mutually-exclusive render branch of the <c>ChangesPanel</c> surface — the native union of the
/// branches the web component renders
/// (web/src/features/admin/components/feature-flags/ChangesPanel.tsx). The web source is a pure
/// presentational component (it takes <c>rows</c>, <c>loading</c> and <c>scopedKey</c> as props and performs
/// no fetching), so the branches are a direct function of the input <see cref="ChangesPanelModel"/>; there is
/// no fetch-driven error / stale / offline branch to reproduce. Every branch maps onto a visible surface —
/// none is ever hidden.
/// </summary>
public enum ChangesPanelState
{
    /// <summary>
    /// The audit log is still loading and no rows are available yet (web <c>loading &amp;&amp; rows == 0</c>)
    /// — the table renders its "Loading audit log…" empty message.
    /// </summary>
    Loading,

    /// <summary>
    /// The audit log resolved with no rows (web <c>!loading &amp;&amp; rows.length === 0</c>) — a friendly
    /// empty state with the scoped or global guidance message.
    /// </summary>
    Empty,

    /// <summary>The audit log produced rows (web fall-through, <c>rows.length &gt; 0</c>) — the table.</summary>
    Data,
}

/// <summary>
/// One inbound flag-change row — the native mirror of the web <c>FeatureFlagChange</c> shape
/// (web/src/types/admin-diagnostics.ts), reduced to the fields the panel renders. <see cref="Id"/> is the
/// stable key the table keys on; <see cref="OldValueJson"/> / <see cref="NewValueJson"/> carry the already
/// <c>JSON.stringify</c>-ed value (or <c>null</c> when the source value was null/absent) so the projection can
/// reproduce the web <c>compact()</c> truncation. Pure data — no WinUI types.
/// </summary>
public sealed record FeatureFlagChangeRow(
    string Id,
    string? ChangedAt,
    string? Actor,
    string FlagKey,
    string Operation,
    string? OldValueJson,
    string? NewValueJson,
    string? Reason);

/// <summary>
/// The render-time data model the <c>ChangesPanel</c> view binds to — the native analogue of the web
/// <c>ChangesPanelProps</c>. The component is presentational: this model carries the <see cref="Rows"/>, the
/// <see cref="Loading"/> flag and the optional <see cref="ScopedKey"/> (the flag key the surface is scoped to,
/// which selects the scoped vs. global empty message). User-facing labels are resolved from the i18n facade by
/// the projection, not passed in. Pure data — no WinUI types — so the projection is unit-tested without a UI
/// host.
/// </summary>
public sealed record ChangesPanelModel(
    IReadOnlyList<FeatureFlagChangeRow> Rows,
    bool Loading,
    string? ScopedKey)
{
    /// <summary>The initial empty model — resolved with no rows and no active scope.</summary>
    public static ChangesPanelModel Empty { get; } =
        new(Array.Empty<FeatureFlagChangeRow>(), false, null);
}

/// <summary>
/// A declarative table column descriptor (key + localized header) — the native, WinUI-free analogue of the web
/// <c>Column&lt;FeatureFlagChange&gt;</c> the panel declares. The view maps each one onto a <c>TsDataColumn</c>;
/// rows address their cells by the same <see cref="Key"/>.
/// </summary>
public sealed record ChangesPanelColumn(string Key, string Header);

/// <summary>
/// A single projected, display-ready table row — the cell values keyed by column key, the stable
/// <see cref="RowKey"/>, the semantic <see cref="OperationStatus"/> the web renders the operation
/// <c>Badge</c> with (preserved here even though the shared <c>TsDataTable</c> renders the operation as text),
/// and a Narrator automation name. Pure data so the projection is unit-tested without a UI host.
/// </summary>
public sealed record ChangesPanelRow(
    string RowKey,
    IReadOnlyDictionary<string, string> Cells,
    StatusKind OperationStatus,
    string AutomationName);

/// <summary>
/// The fully projected, render-ready view of the panel for one input model — the native analogue of the branch
/// the web <c>ChangesPanel</c> returns. Holds the resolved empty/loading labels, the active <see cref="State"/>,
/// and the table columns + rows. Pure data so every branch is asserted headlessly.
/// </summary>
public sealed record ChangesPanelDisplay(
    ChangesPanelState State,
    string EmptyTitle,
    string EmptyMessage,
    string LoadingMessage,
    IReadOnlyList<ChangesPanelColumn> Columns,
    IReadOnlyList<ChangesPanelRow> Rows,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="ChangesPanelModel"/> to its <see cref="ChangesPanelDisplay"/> — the native
/// port of the branch selection in web/src/features/admin/components/feature-flags/ChangesPanel.tsx. The branch
/// precedence mirrors the web source exactly: a non-empty row set renders the table (data); an empty set renders
/// the table's "Loading audit log…" message while loading, otherwise the friendly empty state. The empty message
/// reproduces the web <c>scopedKey ? scopedMessage : globalMessage</c> choice, the value columns reproduce the
/// web <c>compact()</c> truncation, the operation column reproduces the web <c>OP_VARIANT</c> mapping, and
/// timestamps render through <see cref="DateTimeFormatting"/> (so <c>now</c> is injected for determinism). Every
/// label resolves through the i18n facade using the same keys the web source feeds into <c>t()</c>. No WinUI
/// types — unit-tested without a UI host.
/// </summary>
public static class ChangesPanelProjection
{
    /// <summary>Column key for the change timestamp column (web <c>key: 'changed_at'</c>).</summary>
    public const string ChangedAtKey = "changed_at";

    /// <summary>Column key for the actor column (web <c>key: 'actor'</c>).</summary>
    public const string ActorKey = "actor";

    /// <summary>Column key for the flag-key column (web <c>key: 'flag_key'</c>).</summary>
    public const string FlagKeyKey = "flag_key";

    /// <summary>Column key for the operation column (web <c>key: 'operation'</c>).</summary>
    public const string OperationKey = "operation";

    /// <summary>Column key for the old-value column (web <c>key: 'old_value'</c>).</summary>
    public const string OldValueKey = "old_value";

    /// <summary>Column key for the new-value column (web <c>key: 'new_value'</c>).</summary>
    public const string NewValueKey = "new_value";

    /// <summary>Column key for the reason column (web <c>key: 'reason'</c>).</summary>
    public const string ReasonKey = "reason";

    /// <summary>Page size for the data table (web <c>pagination.defaultPageSize</c>).</summary>
    public const int PageSize = 25;

    private const string EmDash = "\u2014";
    private const string Ellipsis = "\u2026";
    private const string SetOperation = "set";
    private const string DeleteOperation = "delete";

    // Web compact(): JSON.stringify, then `s.length > 60 ? s.slice(0, 57) + '…'`.
    private const int MaxValueLength = 60;
    private const int TruncatedLength = 57;

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web props).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    /// <param name="now">The reference instant for timestamp formatting.</param>
    public static ChangesPanelDisplay Project(
        ChangesPanelModel model,
        ILocalizer localizer,
        DateTimeOffset now)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        string emptyTitle = localizer.GetString("admin.flags.audit.empty.title", "No flag changes yet");
        string loadingMessage = localizer.GetString("admin.flags.audit.loading", "Loading audit log\u2026");
        string emptyMessage = BuildEmptyMessage(model.ScopedKey, localizer);

        IReadOnlyList<ChangesPanelColumn> columns = BuildColumns(localizer);
        IReadOnlyList<ChangesPanelRow> rows = BuildRows(model.Rows, now);
        ChangesPanelState state = SelectState(model);

        return new ChangesPanelDisplay(
            State: state,
            EmptyTitle: emptyTitle,
            EmptyMessage: emptyMessage,
            LoadingMessage: loadingMessage,
            Columns: columns,
            Rows: rows,
            AutomationName: BuildAutomationName(state, emptyTitle, emptyMessage, loadingMessage, rows.Count));
    }

    /// <summary>
    /// Branch selection from the web source: a non-empty row set is the data table; an empty set is the
    /// loading message while <c>loading</c>, otherwise the friendly empty state.
    /// </summary>
    private static ChangesPanelState SelectState(ChangesPanelModel model)
    {
        if (model.Rows.Count > 0)
        {
            return ChangesPanelState.Data;
        }

        return model.Loading ? ChangesPanelState.Loading : ChangesPanelState.Empty;
    }

    // Web parity: `scopedKey ? t('…scopedMessage', { key }) : t('…globalMessage')`. An empty scope is falsy.
    private static string BuildEmptyMessage(string? scopedKey, ILocalizer localizer)
    {
        if (string.IsNullOrEmpty(scopedKey))
        {
            return localizer.GetString(
                "admin.flags.audit.empty.globalMessage",
                "Flag changes will appear here once an operator edits a value.");
        }

        string template = localizer.GetString(
            "admin.flags.audit.empty.scopedMessage",
            "No audit rows for \"{0}\" \u2014 edit the value above to start the trail.");
        return string.Format(CultureInfo.CurrentCulture, template, scopedKey);
    }

    private static IReadOnlyList<ChangesPanelColumn> BuildColumns(ILocalizer localizer) =>
    [
        new ChangesPanelColumn(ChangedAtKey, localizer.GetString("admin.flags.audit.cols.changedAt", "Changed at")),
        new ChangesPanelColumn(ActorKey, localizer.GetString("admin.flags.audit.cols.actor", "Actor")),
        new ChangesPanelColumn(FlagKeyKey, localizer.GetString("admin.flags.audit.cols.flagKey", "Key")),
        new ChangesPanelColumn(OperationKey, localizer.GetString("admin.flags.audit.cols.operation", "Op")),
        new ChangesPanelColumn(OldValueKey, localizer.GetString("admin.flags.audit.cols.oldValue", "Old")),
        new ChangesPanelColumn(NewValueKey, localizer.GetString("admin.flags.audit.cols.newValue", "New")),
        new ChangesPanelColumn(ReasonKey, localizer.GetString("admin.flags.audit.cols.reason", "Reason")),
    ];

    private static IReadOnlyList<ChangesPanelRow> BuildRows(
        IReadOnlyList<FeatureFlagChangeRow> rows,
        DateTimeOffset now)
    {
        if (rows.Count == 0)
        {
            return Array.Empty<ChangesPanelRow>();
        }

        var projected = new List<ChangesPanelRow>(rows.Count);
        foreach (var row in rows)
        {
            string changedAt = FormatChangedAt(row.ChangedAt, now);
            string actor = string.IsNullOrEmpty(row.Actor) ? EmDash : row.Actor;
            string flagKey = row.FlagKey; // Web renders row.flag_key verbatim (no fallback).
            string operation = row.Operation;
            string oldValue = Compact(row.OldValueJson);
            string newValue = Compact(row.NewValueJson);
            string reason = string.IsNullOrEmpty(row.Reason) ? EmDash : row.Reason;

            var cells = new Dictionary<string, string>(StringComparer.Ordinal)
            {
                [ChangedAtKey] = changedAt,
                [ActorKey] = actor,
                [FlagKeyKey] = flagKey,
                [OperationKey] = operation,
                [OldValueKey] = oldValue,
                [NewValueKey] = newValue,
                [ReasonKey] = reason,
            };

            projected.Add(new ChangesPanelRow(
                RowKey: row.Id,
                Cells: cells,
                OperationStatus: OperationStatusFor(operation),
                AutomationName: BuildRowAutomationName(changedAt, flagKey, operation, actor, oldValue, newValue, reason)));
        }

        return projected;
    }

    // Web parity for the operation Badge: OP_VARIANT[op] (set→success, delete→danger) ?? neutral.
    private static StatusKind OperationStatusFor(string operation) => operation switch
    {
        SetOperation => StatusKind.Success,
        DeleteOperation => StatusKind.Danger,
        _ => StatusKind.Neutral,
    };

    // Web parity for the changed_at column: <TimeStamp value format="absolute" /> → the locale-aware
    // absolute datetime, with the universal em-dash fallback for null / unparseable timestamps.
    private static string FormatChangedAt(string? raw, DateTimeOffset now)
    {
        if (TryParseTimestamp(raw, out var value))
        {
            return DateTimeFormatting.Format(value, DateTimeVariant.Full, now);
        }

        return EmDash;
    }

    private static bool TryParseTimestamp(string? raw, out DateTimeOffset value)
    {
        if (!string.IsNullOrWhiteSpace(raw) && DateTimeOffset.TryParse(
                raw,
                CultureInfo.InvariantCulture,
                DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                out value))
        {
            return true;
        }

        value = default;
        return false;
    }

    // Web compact(): `value == null` → '—'; otherwise JSON.stringify and, when longer than 60 chars,
    // slice(0, 57) + '…'. The model already carries the stringified value (or null for a null/absent value).
    internal static string Compact(string? json)
    {
        if (string.IsNullOrEmpty(json))
        {
            return EmDash;
        }

        return json.Length > MaxValueLength
            ? string.Concat(json.AsSpan(0, TruncatedLength), Ellipsis)
            : json;
    }

    private static string BuildRowAutomationName(
        string changedAt,
        string flagKey,
        string operation,
        string actor,
        string oldValue,
        string newValue,
        string reason) =>
        $"{changedAt}. {flagKey}. {operation}. {actor}. {oldValue} {newValue}. {reason}";

    private static string BuildAutomationName(
        ChangesPanelState state,
        string emptyTitle,
        string emptyMessage,
        string loadingMessage,
        int rowCount) => state switch
        {
            ChangesPanelState.Empty => $"{emptyTitle}. {emptyMessage}",
            ChangesPanelState.Loading => loadingMessage,
            _ => string.Create(CultureInfo.InvariantCulture, $"{emptyTitle}. {rowCount}"),
        };
}

/// <summary>
/// PII-safe diagnostics for the <c>ChangesPanel</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never an actor, flag key, value or reason — so a
/// diagnostics line can never leak which operator or flag was involved. Thread-safe.
/// </summary>
public sealed class ChangesPanelDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public ChangesPanelDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=ChangesPanel</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={ChangesPanelRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>ChangesPanel</c> feature surface — the native mirror of the web component at
/// <c>web/src/features/admin/components/feature-flags/ChangesPanel.tsx</c>.
/// </summary>
public static class ChangesPanelRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "ChangesPanel";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/ChangesPanel/ChangesPanelModel.kt
// Pure, framework-free model + projection for the ChangesPanel feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/admin/components/feature-flags/ChangesPanel.tsx). No Compose, no Android, no HTTP:
// every declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// The web component is purely presentational — its parent (the Feature-Flags page, `useFlagChanges`) loads
// the `FeatureFlagChange[]` and passes it down with a `loading` flag and an optional `scopedKey`. This file
// owns the parts the web column renderers + `compact()` helper compute from those props: the operation →
// semantic-tone classification (web `OP_VARIANT`), the actor/reason "—" fallbacks (web `value || '—'`), the
// `compact()` JSON preview (stringify, 60-char cap, "…" suffix, null → "—"), and the `changed_at` timestamp
// formatting (web `<TimeStamp format="absolute" />`). It binds the shared P1/S8 `FeatureFlagChange` wire
// model verbatim — no re-declaration — so the cached payload flows straight through.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/ChangesPanel — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting types.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.changespanel

import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.presentation.featureflags.FeatureFlagChange
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import java.time.Instant
import java.time.LocalDateTime
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.time.format.FormatStyle
import java.util.Locale

/** Em dash shown for a null value or a blank actor/reason — the web `'—'` fallback. */
internal const val EM_DASH: String = "\u2014"

/** Single-character horizontal ellipsis appended to a truncated value preview — the web `'…'`. */
internal const val ELLIPSIS: String = "\u2026"

/** Web `compact()` truncation gate: stringified values longer than this are clipped. */
internal const val COMPACT_MAX_LENGTH: Int = 60

/** Web `compact()` retained-prefix length before the ellipsis (`s.slice(0, 57) + '…'`). */
internal const val COMPACT_KEEP_LENGTH: Int = 57

/** Compact (non-pretty) JSON, matching the web `JSON.stringify(value)` preview with no spacing. */
private val COMPACT_JSON: Json = Json.Default

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object ChangesPanelRegistration {
    /** Stable surface id. */
    const val ID: String = "changes-panel"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11); carries no flag payload. */
    const val SLUG: String = "ChangesPanel"
}

/**
 * The semantic tone of a flag-change operation — the native analogue of the web
 * `OP_VARIANT: Record<FeatureFlagOperation, 'success' | 'danger'>` plus its `?? 'neutral'` fallback. The
 * composable maps each case to a [io.teslasync.android.components.ui.BadgeVariant]; the raw operation string
 * remains the badge label (web renders `{row.operation}`).
 */
enum class OperationTone {
    /** A `set` write — web `success`. */
    Positive,

    /** A `delete` write — web `danger`. */
    Negative,

    /** Any other/unknown operation — web `?? 'neutral'`. */
    Neutral,
}

/**
 * One fully projected, render-ready audit row — the native analogue of a single web `DataTable` row after
 * its per-column renderers have run. Pure data (no Compose types) so the projection is unit-tested without a
 * UI host: [changedAt] is already formatted by the caller's locale/zone, [oldValue]/[newValue] are already
 * `compact()`-ed, [actor]/[reason] already carry the "—" fallback, and [tone] classifies [operation] for the
 * badge. [id] is the stable row key (web `keyExtractor={(row) => row.id}`).
 */
@Suppress("LongParameterList") // A render-ready row mirrors the web table's seven columns plus its key.
data class FlagChangeRow(
    val id: Long,
    val changedAt: String,
    val actor: String,
    val flagKey: String,
    val operation: String,
    val tone: OperationTone,
    val oldValue: String,
    val newValue: String,
    val reason: String,
)

/**
 * Already-localized column-header microcopy — the web `t('admin.flags.audit.cols.*')` keys. The composable
 * resolves these through the P1/S10 i18n facade (`stringResource`) and passes them in; tests pass a
 * deterministic instance, keeping the surface free of any English literal.
 */
data class ChangesPanelColumnLabels(
    val changedAt: String,
    val actor: String,
    val flagKey: String,
    val operation: String,
    val oldValue: String,
    val newValue: String,
    val reason: String,
)

/**
 * Already-localized empty/loading microcopy — the web `t('admin.flags.audit.empty.*')` /
 * `t('admin.flags.audit.loading')` keys. [scopedMessage] is a lambda so the composable resolves the `%1$s`
 * flag-key argument through `Context.getString` (web `{{key}}` interpolation); tests pass a deterministic
 * one.
 */
data class ChangesPanelEmptyLabels(
    val title: String,
    val globalMessage: String,
    val loadingLabel: String,
    val scopedMessage: (key: String) -> String,
)

/**
 * The localized microcopy the surface renders — the column headers plus the empty/loading strings. The
 * composable builds this from `stringResource`/`Context.getString`; tests pass a deterministic instance.
 */
data class ChangesPanelStrings(
    val columns: ChangesPanelColumnLabels,
    val empty: ChangesPanelEmptyLabels,
)

/**
 * The pure projection the composable renders — the native mirror of the web component's per-column render
 * functions and `compact()` helper. Stateless and side-effect-free so it is fully covered by the off-device
 * unit gate.
 */
object ChangesPanelProjection {
    /**
     * Classifies an [operation] into its badge [OperationTone] — web `OP_VARIANT[op] ?? 'neutral'`: `set` is
     * positive (success), `delete` is negative (danger), anything else folds to neutral.
     */
    fun toneFor(operation: String): OperationTone =
        when (operation) {
            "set" -> OperationTone.Positive
            "delete" -> OperationTone.Negative
            else -> OperationTone.Neutral
        }

    /** A non-empty value, else the em-dash fallback — web `value || '—'` (JS treats `""` as falsy). */
    fun orDash(value: String): String = value.ifEmpty { EM_DASH }

    /**
     * The compact one-line JSON preview of a flag value — the native port of the web `compact(value)`: a JSON
     * `null` (kotlinx [JsonNull]) renders the em-dash (web `value == null`), otherwise the value is stringified
     * (web `JSON.stringify`) and, when longer than [COMPACT_MAX_LENGTH], clipped to [COMPACT_KEEP_LENGTH]
     * characters plus the [ELLIPSIS] (web `s.slice(0, 57) + '…'`).
     */
    fun compact(value: JsonElement): String {
        if (value is JsonNull) return EM_DASH
        val serialized = COMPACT_JSON.encodeToString(JsonElement.serializer(), value)
        return if (serialized.length > COMPACT_MAX_LENGTH) {
            serialized.substring(0, COMPACT_KEEP_LENGTH) + ELLIPSIS
        } else {
            serialized
        }
    }

    /**
     * Projects the loaded [rows] into render-ready [FlagChangeRow]s, preserving order. Each field is mapped
     * exactly as the matching web column renderer does — the actor/reason "—" fallback, the operation tone,
     * the `compact()` value previews — and [formatTime] formats `changed_at` (injecting it keeps this function
     * locale/zone-deterministic for tests; the composable supplies the real localized formatter). An empty
     * input yields no rows so the composable shows the empty state.
     */
    fun project(
        rows: List<FeatureFlagChange>,
        formatTime: (changedAt: String) -> String,
    ): List<FlagChangeRow> =
        rows.map { row ->
            FlagChangeRow(
                id = row.id,
                changedAt = formatTime(row.changedAt),
                actor = orDash(row.actor),
                flagKey = row.flagKey,
                operation = row.operation,
                tone = toneFor(row.operation),
                oldValue = compact(row.oldValue),
                newValue = compact(row.newValue),
                reason = orDash(row.reason),
            )
        }
}

/**
 * Tolerant ISO-8601 → localized "medium date, short time" formatter — the native analogue of the web
 * `<TimeStamp format="absolute" />` (`toLocaleString` with `{year, month:'short', day, hour, minute}`). Pure
 * (java.time only) so it is unit-tested deterministically with a fixed zone/locale. A blank or unparseable
 * input yields [EM_DASH], exactly like the web component's invalid-date guard.
 */
object ChangesPanelTimeFormatting {
    fun format(
        changedAt: String,
        zone: ZoneId,
        locale: Locale,
    ): String {
        val instant = parseInstant(changedAt) ?: return EM_DASH
        return DateTimeFormatter
            .ofLocalizedDateTime(FormatStyle.MEDIUM, FormatStyle.SHORT)
            .withLocale(locale)
            .withZone(zone)
            .format(instant)
    }

    // Tolerant decode chain: an RFC-3339 instant ("…Z"), then an offset date-time, then a zoneless local
    // date-time treated as UTC. The first that parses wins; none parsing yields the em-dash guard above.
    private val parsers: List<(String) -> Instant?> =
        listOf(
            { raw -> tryParse { Instant.parse(raw) } },
            { raw -> tryParse { OffsetDateTime.parse(raw).toInstant() } },
            { raw -> tryParse { LocalDateTime.parse(raw).toInstant(ZoneOffset.UTC) } },
        )

    private fun parseInstant(raw: String): Instant? = if (raw.isBlank()) null else parsers.firstNotNullOfOrNull { it(raw) }

    private fun tryParse(block: () -> Instant): Instant? =
        try {
            block()
        } catch (_: DateTimeParseException) {
            null
        }
}

/**
 * Emits the one PII-safe `view.opened` diagnostic with the surface [ChangesPanelRegistration.SLUG] (P1/S11).
 * Kept free of Compose so it is unit-tested with a recording [Logger]; the composable calls it from its
 * first-composition effect.
 */
fun recordChangesPanelOpened(logger: Logger) {
    logger.info("view.opened", mapOf("surface" to ChangesPanelRegistration.SLUG))
}
